### PR TITLE
DM-29522 Add spatially varying zogy

### PR DESCRIPTION
**Pull request by kgabor**
* Add grid generation method. Ensure outer box is always even.
 * Add border initialization by random noise method.
 * Add exposure split and paste back methods.
 * Add spatial psf construction by coaddPsf.
 * Set calib property to the score image as well.
 * Update some log messages to use f-methods to avoid unnecessary string formatting.
 * Remove unused code parts.
 * Add subexp variance robustness.

### CODE REVIEW

File: tests/test_zogy.py [hunk 1]
```diff
@@ -259,6 +268,37 @@ def testFourierTransformConvention(self):
...
+        innerbox = bb.erodedBy(geom.Extent2I(3, 4))
+        D[innerbox] = 1
+
+        borderboxes = task.splitBorder(innerbox, bb)
+        for x in borderboxes:
+            D[x] += 1
+        # The splitting should cover all border pixels exactly once
+        self.assertTrue(np.all(D.array == 1), "Border does not cover all pixels exactly once.")
+
+    def testGenerateGrid(self):
```
Comment (morriscb) [inline tests/test_zogy.py:287]: Is there possibly a need to test the case where the "grid" has only one cell? That is the full exposures are used and not broken up?

**Review by morriscb — APPROVED** [review]
I think overall it looks good. I did find one instance where exposure1 was used instead of exposure2 as intended (I think) so likely take a look through the code again to make sure you have all of those variables pointing to the right place.

File: python/lsst/ip/diffim/zogy.py [hunk 2]
```diff
@@ -377,6 +641,12 @@ def prepareFullExposure(self, exposure1, exposure2, correctBackground=False):
 
         exposure1 = exposure1.clone()
         exposure2 = exposure2.clone()
+        # Fallback values if sub exposure variance calculation is problematic
+        sig1 = np.sqrt(self._computeVarianceMean(exposure1))
+        self.fullExpVar1 = sig1*sig1
+        sig2 = np.sqrt(self._computeVarianceMean(exposure1))
```
[the remark this answers is not in the document]
Comment (kgabor) [reply python/lsst/ip/diffim/zogy.py:647]: Good catch.

File: python/lsst/ip/diffim/zogy.py [hunk 3]
```diff
@@ -193,38 +207,261 @@ def padCenterOriginArray(A, newShape, useInverse=False, dtype=None):
...
+
+        Returns
+        -------
+        boxList : `list` of `lsst.pipe.base.Struct`
+          ``innerBox``, ``outerBox`` : `lsst.geom.Box2I`, inner boxes and overlapping border around them.
+
+        """
+        powersOf2 = np.array([16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192])
+        doubleEdgeDims = minEdgeDims * 2
+        xsize, ysize = imageBox.getDimensions()
```
[the remark this answers is not in the document]
Comment (kgabor) [reply python/lsst/ip/diffim/zogy.py:378]: Indeed, there is a bit of terminology confusion here...

File: tests/test_zogy.py (hunk 1 above)
Comment (kgabor) [reply tests/test_zogy.py:287]: Added.

**Pull request merged**